### PR TITLE
feat: implement border text function

### PR DESCRIPTION
**Pull request by romgrk**
Link: https://github.com/MunifTanjim/nui.nvim/pull/142

### CODE REVIEW

**Comment by romgrk** [comment]
> Makes sense that you prefer to have a function here. I'd like that function to return string | string[] | NuiText | NuiText[] instead of NuiLine (which represents a whole Line in a buffer). And mid_char should be passed as 2nd param to that function.

I've added the `mid_char` parameter.

For the polymorphic return type, is it really necessary? For the simple use-cases, the user already has access to `border:set_text(edge, text)`. Making the rendering function have multiple return types will complicate the code without offering a clear advantage, imho.

**Comment by MunifTanjim** [comment]
> For the polymorphic return type, is it really necessary?

> Making the rendering function have multiple return types will complicate the code without offering a clear advantage, imho.

Hmm... Yep, you are right. Let's leave that polymorphic return from here.

File: lua/nui/popup/border.lua [hunk 1]
```diff
@@ -143,17 +143,20 @@ local function calculate_buf_edge_line(internal, edge, text, align)
 
   local max_width = size.width - left_char:width() - right_char:width()
 
-  local content_text = Text(defaults(text, ""))
-  if mid_char:width() == 0 then
-    content_text:set(string.rep(" ", max_width))
+  local content_line
+  if type(text) == "function" then
+    content_line = text(max_width, mid_char)
```
Comment (MunifTanjim) [inline lua/nui/popup/border.lua:148]: need to check the `content_line:width()` here and raise error if it's greater than `max_width`

File: lua/nui/popup/border.lua [hunk 2]
```diff
@@ -541,7 +544,7 @@ function Border:_relayout()
 end
 
 ---@param edge "'top'" | "'bottom'"
----@param text? nil | string | table # string or NuiText
+---@param text? nil | string | table | function # string or NuiText or render(width)
```
Comment (MunifTanjim) [inline lua/nui/popup/border.lua:547]:
```suggestion
---@param text? nil | string | table | function # string or NuiText or render(max_width, mid_char)
```

Also update the README for `NuiLine` and `Popup` about the changes?

**Comment by romgrk** [comment]
Closing, stale.

**Pull request closed without merging**